src/shogun/lib/type_case.h

```diff
@@ -0,0 +1,152 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#ifndef SHOGUN_TYPE_CASE_H
+#define SHOGUN_TYPE_CASE_H
+
+#include <typeindex>
+#include <unordered_map>
+
+#include <shogun/lib/any.h>
+#include <shogun/lib/type_list.h>
+
+using namespace shogun;
+
+// here we define a list of all types that are implemented in shogun
+// the head of the list returns the type
+typedef Types<
+    bool, char, int8_t, uint8_t, int16_t, uint16_t, int32_t, uint32_t, int64_t,
+    uint64_t, float32_t, float64_t, floatmax_t>::type AllTypes;
```

Comment:
sg_types
SG_TYPES
?
Also, those are only scalar types. What about adding a matrix type to start seeing what happens in those cases?
We will also need std::vector and friends.....which might show the issues with this approach: combinatorial explosions. 
I guess we could argue that the number of types are limited.